File: tools/src/main/scala/com/threerings/nexus/streamergen/Scanner.scala
//
// $Id$

package com.threerings.nexus.streamergen;

import scala.collection.JavaConversions._

import javax.annotation.processing.{Messager, ProcessingEnvironment}

import javax.lang.model.element.{Element, ElementKind, ExecutableElement, Modifier}
import javax.lang.model.element.{Name, TypeElement, TypeParameterElement, VariableElement}
import javax.lang.model.`type`.{DeclaredType, TypeKind, TypeMirror, TypeVariable}
import javax.lang.model.util.{ElementScanner6, Types}

/**
 * Does the actual AST walking and computes metadata that is used to generate {@code Streamer}
 * source files.
 */
class Scanner (env :ProcessingEnvironment) extends ElementScanner6[Unit, Unit]
{
  /**
   * Scans the supplied compilation unit and returns metadata for all encountered classes, mapped
   * by name.
   */
  def scanUnit (e :Element) :Seq[ClassMetadata] = {
    try {
      scan(e)
      _metas
    } finally {
      _metas = Seq()
    }
  }

  override def visitType (e :TypeElement, p :Unit) {
    val meta = new ClassMetadata(e.getQualifiedName, e.getSimpleName, e.getSuperclass,
                                 getEncloser(e).getQualifiedName,
                                 e.getModifiers.contains(Modifier.ABSTRACT))
    e.getTypeParameters foreach { tpe => meta.typeParams.put(tpe.getSimpleName, tpe.asType) }
    // only add this class if it's streamable, but let it be processed regardless, as it may
    // contain nested types which are themselves streamable
    if (isStreamable(e)) _metas = _metas :+ meta
    _mstack = meta :: _mstack
    super.visitType(e, p)
    _mstack = _mstack.tail
    // System.err.println("Processed type " + e.getSimpleName + " -> " + meta)
  }

  override def visitExecutable (e :ExecutableElement, p :Unit) {
    super.visitExecutable(e, p)
    if (e.getKind == ElementKind.CONSTRUCTOR) {
      e.getParameters foreach { arg => _mstack.head.ctorArgs.put(arg.getSimpleName, arg.asType) }
    }
  }

  override def visitVariable (e :VariableElement, p :Unit) {
    super.visitVariable(e, p)
    if (e.getKind == ElementKind.FIELD) {
      // System.err.println("Noting field " + e)
      _mstack.head.fields.put(e.getSimpleName, e.asType)
    }
  }

  protected def isStreamable (e :TypeElement) :Boolean = {
    e.getInterfaces.exists(isStreamableIfc) || (e.getSuperclass match {
      case dt :DeclaredType => isStreamable(dt.asElement.asInstanceOf[TypeElement])
      case _ => false
    })
  }

  protected def isStreamableIfc (t :TypeMirror) = t match {
    case dt :DeclaredType => dt.asElement.asInstanceOf[TypeElement].getQualifiedName.toString ==
      "com.threerings.nexus.io.Streamable"
    case _ => false
  }

  protected def getEncloser (e :TypeElement) :TypeElement = e.getEnclosingElement match {
    case ee :TypeElement if (ee != null) => getEncloser(ee)
    case _ => e
  }

  protected val _msgr = env.getMessager
  protected val _types = env.getTypeUtils

  protected var _metas = Seq[ClassMetadata]()
  protected var _mstack :List[ClassMetadata] = Nil
}

File: tools/src/main/java/com/threerings/nexus/streamergen/Scanner.java
//
// $Id$

package com.threerings.nexus.streamergen;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;

import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.Name;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.TypeParameterElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.type.TypeVariable;
import javax.lang.model.util.ElementScanner6;
import javax.lang.model.util.Types;

import com.google.common.collect.Maps;

/**
 * Does the actual AST walking and computes metadata that is used to generate {@code Streamer}
 * source files.
 */
public class Scanner extends ElementScanner6<Void, Void>
{
    public Scanner (ProcessingEnvironment env) {
        _msgr = env.getMessager();
        _types = env.getTypeUtils();
    }

    /**
     * Scans the supplied compilation unit and returns metadata for all encountered classes, mapped
     * by name.
     */
    public Map<Name,ClassMetadata> scanUnit (Element e) {
        Map<Name,ClassMetadata> result = (_map = Maps.newHashMap());
        scan(e);
        return result;
    }

    @Override
    public Void visitType (TypeElement e, Void p) {
        ClassMetadata meta = new ClassMetadata(e.getSuperclass(),
                                               e.getModifiers().contains(Modifier.ABSTRACT));
        for (TypeParameterElement tpe : e.getTypeParameters()) {
            meta.typeParams.put(tpe.getSimpleName(), tpe.asType());
        }
        _map.put(e.getSimpleName(), meta);
        _metas.push(meta);
        super.visitType(e, p);
        _metas.pop();
        System.err.println("Processed type " + e.getSimpleName() + " -> " + meta);
        return null;
    }

    @Override
    public Void visitExecutable (ExecutableElement e, Void p) {
        super.visitExecutable(e, p);
        if (e.getKind() == ElementKind.CONSTRUCTOR) {
            for (VariableElement arg : e.getParameters()) {
                _metas.peek().ctorArgs.put(arg.getSimpleName(), arg.asType());
            }
        }
        return null;
    }

    @Override
    public Void visitVariable (VariableElement e, Void p) {
        super.visitVariable(e, p);
        if (e.getKind() == ElementKind.FIELD) {
            System.err.println("Noting field " + e);
            _metas.peek().fields.put(e.getSimpleName(), e.asType());
        }
        return null;
    }

    protected final Messager _msgr;
    protected final Types _types;

    protected Deque<ClassMetadata> _metas = new ArrayDeque<ClassMetadata>();
    protected Map<Name,ClassMetadata> _map;
}
